net/speedtest/speedtest.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package speedtest contains both server and client code for
+// running speedtests between tailscale nodes.
+package speedtest
+
+import (
+	"fmt"
+	"time"
+)
+
+const (
+	LenBufData        = 32000 // Size of the block of randomly generated data to send.
+	DefaultNumSeconds = 5     // Default time for a test
+	MaxNumSeconds     = 30
+	Version           = 1
+)
+
+// This is the initial message sent to the sever, that contains information on how to
+// conduct the test.
+type TestConfig struct {
+	Version      int           `json:"version"`
+	Type         string        `json:"type"`
+	TestDuration time.Duration `json:"time"`
+	Increment    int           `json:"inc"`
+}
+
+// This is the response to the TestConfig message. It contains errors that the server might
+// have with the TestConfig.
+type TestConfigResponse struct {
+	Errors string `json:"errors,omitempty"`
+}
+
+// This represents the Result of a speedtest within a specific interval
+type Result struct {
+	startTime  float64
+	endTime    float64
+	mbRecieved float64
+	secPassed  float64
```

Comment:
Having both {startTime, endTime} and secPassed makes us have to consider what we'd do if they don't match. I think have separate start and end times doesn't provide useful information versus just secPassed, recommend having only secPassed.